Modules/Common/CardanoLikeNativeTokensModule.php

```diff
@@ -0,0 +1,321 @@
+    }
+
+    final public function pre_process_block($block_id)
+    {
+        $events = [];
+        $block = [];
+
+        $transactions = pg_fetch_all(pg_query_params($this->db,
+            'SELECT id, hash, out_sum, fee, deposit FROM tx WHERE block_id = $1 ORDER BY block_index', [$this->block_db_id]));
+        // or this can be done via a subquery like this: WHERE block_id = (SELECT id FROM block WHERE block_no = $1),
+        // and $block_id is the param
+
+        if (count($transactions) !== $this->transaction_count)
+            throw new ConsensusException('Transaction count mismatch (orphaned block?)');
+
+        if (!$transactions)
+        {
+            $this->set_return_events([]);
+            $this->set_return_currencies([]);
+            return;
+        }
+
+        $in_query = [];
+
+        foreach ($transactions as $transaction)
+        {
+            $block[($transaction['id'])] = $transaction;
+            $block[($transaction['id'])]['inputs'] = [];
+            $block[($transaction['id'])]['outputs'] = [];
+            $in_query[] = $transaction['id'];
+        }
+
+        $in_query = implode(', ', $in_query);
+
+        $outs = pg_fetch_all(pg_query($this->db,
+            "SELECT tx_out.tx_id, tx_out.address, multi_asset.fingerprint, ma_tx_out.ident AS token, ma_tx_out.quantity
+                    FROM tx_out
+                    JOIN ma_tx_out ON (ma_tx_out.tx_out_id = tx_out.id)
+                    JOIN multi_asset ON ma_tx_out.ident = multi_asset.id
+                    WHERE tx_out.tx_id in ({$in_query})
+                    ORDER BY tx_out.tx_id, tx_out.index"));
+
+        foreach ($outs as $out)
+        {
+            $block[($out['tx_id'])]['outputs'][] = $out;
+        }
+
+        // inputs - due to the way foreign keys are chosen in relevant tables, 3-table-join ends up 8-10 times slower then this:
+
+        $ins_1 = pg_fetch_all(pg_query($this->db,
+            "SELECT tx_in.tx_in_id, tx_out.address, tx_in.id as orderby, tx_out.id as nextkey
+                    FROM tx_in
+                    JOIN tx_out ON (tx_in.tx_out_id = tx_out.tx_id AND tx_in.tx_out_index = tx_out.index)
+                    WHERE tx_in.tx_in_id IN ({$in_query})"));
+
+        if (count($ins_1) < 1)
+        {
+            $this->set_return_events([]);
+            $this->set_return_currencies([]);
+            return;
+        }
+
+        $ins_dict = [];
+        $subquery = [];
+
+        foreach ($ins_1 as $aux)
+        {
+                $ins_dict[$aux['nextkey']] = $aux;
+                $subquery[] = $aux['nextkey'];
+        }
+        $subquery = implode(', ', $subquery);
+
+        $ins_2 = pg_fetch_all(pg_query($this->db,
+            "SELECT ma_tx_out.tx_out_id as id, multi_asset.fingerprint, ident AS token, ma_tx_out.quantity
+                    FROM ma_tx_out
+                    JOIN multi_asset ON ma_tx_out.ident = multi_asset.id
+                    WHERE tx_out_id in ({$subquery})"));
```

Comment:
No order

Reply:
LGTM! Ordering is kept in the code, see comment above.